lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -251,85 +215,51 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
                            newestHash
         send conv $ mkBlocksRequest lcaChildHash newestHash
         logDebug "Requested blocks, waiting for the response"
-        chainE <- runExceptT (retrieveBlocks conv lcaChild newestHash)
+        chainE <- runExceptT (retrieveBlocks conv numBlocks)
         case chainE of
             Left e -> do
-                let msg = sformat ("Error retrieving blocks frmo "%shortHashF%
+                let msg = sformat ("Error retrieving blocks from "%shortHashF%
                                    " to "%shortHashF%" from peer "%
                                    build%": "%stext)
                                   lcaChildHash newestHash nodeId e
                 logWarning msg
                 throwM $ DialogUnexpected msg
-            Right blocks -> do
-                logDebug $ sformat
-                    ("Retrieved "%int%" blocks of total size "%builder%": "%buildListBounds)
-                    (blocks ^. _NewestFirst . to NE.length)
-                    (unitBuilder $ biSize blocks)
-                    (getNewestFirst $ map headerHash blocks)
-                return blocks
+            Right bs -> case nonEmptyNewestFirst bs of
+                Nothing -> do
+                    let msg = sformat ("Peer gave an empty blocks list")
+                    throwM $ DialogUnexpected msg
+                Just blocks -> return blocks
 
     -- A piece of the block retrieval conversation in which the blocks are
     -- pulled in one-by-one.
     retrieveBlocks
         :: ConversationActions MsgGetBlocks MsgBlock d
-        -> BlockHeader
-        -> HeaderHash
-        -> ExceptT Text d (NewestFirst NE Block)
-    retrieveBlocks conv lcaChild endH = do
-        blocks <- retrieveBlocksDo 0 conv (lcaChild ^. prevBlockL) endH []
-        let b0 = blocks ^. _NewestFirst . _neHead
-        if headerHash b0 == headerHash lcaChild
-           then pure blocks
-           else throwError $ sformat
-                    ("First block of chain is "%build%
-                     " instead of expected "%build)
-                    (b0 ^. blockHeader) lcaChild
+        -> Int
+        -> ExceptT Text d (NewestFirst [] Block)
+    retrieveBlocks conv numBlocks = retrieveBlocksDo conv numBlocks []
 
     -- Content of retrieveBlocks.
-    -- Receive until a block with a given hash 'endH' is found.
-    -- Assumes that 'endH' is the hash of a block which is older than 'prevH'.
-    -- Retrieval starts with 'prevH'.
+    -- Receive a given number of blocks. If the server doesn't send this
+    -- many blocks, an error will be given.
```

Comment:
Alright, so `getBlocks` won't return empty list if it's asked to get `[a..b]` where `b > a` and `getHeaders` request didn't return garbage.